crates/ide_assists/src/handlers/replace_if_let_with_match.rs

```diff
@@ -235,12 +235,14 @@ fn pick_pattern_and_expr_order(
 ) -> Option<(ast::Pat, ast::Expr, ast::Expr)> {
     let res = match (pat, pat2) {
         (ast::Pat::WildcardPat(_), _) => return None,
-        (pat, sad_pat) if is_sad_pat(sema, &sad_pat) => (pat, expr, expr2),
-        (sad_pat, pat) if is_sad_pat(sema, &sad_pat) => (pat, expr2, expr),
+        (pat, _) if expr2.syntax().first_child().is_none() => (pat, expr, expr2),
+        (_, pat) if expr.syntax().first_child().is_none() => (pat, expr2, expr),
         (pat, pat2) => match (binds_name(&pat), binds_name(&pat2)) {
-            (true, true) => return None,
             (true, false) => (pat, expr, expr2),
             (false, true) => (pat2, expr2, expr),
+            _ if is_sad_pat(sema, &pat2) => (pat, expr, expr2),
+            _ if is_sad_pat(sema, &pat) => (pat2, expr2, expr),
+            (true, true) => return None,
             (false, false) => (pat, expr, expr2),
```

Comment:
The `return None` arm has to go above the sad pat guarded ones. If both arms bind a name it cannot be transformed into an `if let`.

On that note, we can move the sad pat guard into the last arm's body instead, that is something like

```rs
(false, false) => if is_sad_pat(sema, &pat) {
    (pat2, expr2, expr)
} else {
    (pat, expr, expr2)
},
```

Reply:
Semantically, I agree here, `return None` _should_ be above the sad guards.
however, it was below them before, too. For good reason:
```rs
match Some(12) {
  Some(x) => println!("{}", x),
  None => println!("Nothing :("),
}
```
  recognizes `None` as an `IdentPat`, and thus counts it as binding a name. I don't know if this is in fact solveable without having more semantic information about the pattern.
  
  If there is no such semantic data available, one thought that just crossed my mind was that we could apply some heuristic here, such as saying that "if the ident pattern is Uppercase, it probably is an enum variant or constant and should thus be counted as not binding". This would still have the potential to misfire - but _significantly_ less often than it currently does, with `None` being seen as an ident pattern. Alternatively, we could specialcase `None` completely - but that seems less optimal than assuming that users follow very basic naming conventions

Reply:
> recognizes None as an IdentPat, and thus counts it as binding a name. 

Oooh, this is a good catch, I forgot about const patterns here when implementing that...
The proper fix is to pull in semantic information here, we can do this by putting the ident pats `name` into `NameClass::classify` then checking whether the result is a `Definition::ConstReference`.

Reply:
Done!